### PR TITLE
adding vlan tagging for cont ns interface

### PR DESCRIPTION
**Pull request by rkamudhan**
Support VLAN tagging for the interface inside the container network namespace
This PR is related to the issue #82

### CODE REVIEW

**Comment by rosenhouse** [comment]
Sorry about the appveyor CI failure, I configured that integration for this repo before it was ready.  Please ignore that failure.

**Comment by squeed** [comment]
I'm not exactly sure what this does - can you please explain? Also, all PRs that add a config flag need to have a corresponding update in the README for that plugin.

**Comment by rkamudhan** [comment]
@squeed  Current VLAN plugin creates VLAN tagging for the master interface in the host namespace. We have a use case in which application inside a pod needs to connect to the master interface, which is in the container network namespace and connect to the application instances on two separate VLAN. The `isContNs` config flag in the PR is used to identify the master interface in the container network namespace.

Will create PR for the VLAN plugin README file.

**Comment by squeed** [comment]
Ah, I get it now. A few thoughts.
1) Could you change the flag to "masterInContainer" - it makes it clearer
2) Could you refactor the code a bit? Make a common function that creates the tagged vlan interface, then call it appropriately?

Something like
```
if masterInContainer {
    containerNS.Do({
        makeVlanInterface()
    })
} else {
    makeVlanInterface()
}
````

**Comment by rkamudhan** [comment]
@squeed Thanks for the feedback, I will do it.

File: plugins/main/vlan/vlan.go [hunk 1]
```diff
@@ -32,9 +33,10 @@ import (
 
 type NetConf struct {
 	types.NetConf
-	Master string `json:"master"`
-	VlanId int    `json:"vlanId"`
-	MTU    int    `json:"mtu,omitempty"`
+	Master            string `json:"master"`
+	MasterInContainer bool   `json:"masterInContainer"`
```
Comment (rkamudhan) [inline plugins/main/vlan/vlan.go]: Changed the parameter name to "masterInContainer"

File: plugins/main/vlan/vlan.go [hunk 2]
```diff
@@ -60,10 +62,29 @@ func loadConf(bytes []byte) (*NetConf, string, error) {
 
 func createVlan(conf *NetConf, ifName string, netns ns.NetNS) (*current.Interface, error) {
 	vlan := &current.Interface{}
+	var curNetns ns.NetNS
+	var m netlink.Link
+	var err error
+
+	if conf.MasterInContainer != false {
```
Comment (rkamudhan) [inline plugins/main/vlan/vlan.go]: checking the "MasterInContainer" flag and applying container network namespace to the find the master and add the vlan link to it.

**Comment by rkamudhan** [comment]
@squeed Added the changes to the code.

**Comment by rkamudhan** [comment]
@squeed Please let me know, whether the PR is OK?

**Comment by rkamudhan** [comment]
@squeed @dcbw Please let me know, Whether I want to make changes for the PR.

**Comment by jellonek** [comment]
@squeed @dcbw @bboreham ping

**Comment by rkamudhan** [comment]
@squeed  @dcbw @bboreham Could you please review this PR.

File: plugins/main/vlan/vlan.go [hunk 3]
```diff
@@ -60,10 +62,29 @@ func loadConf(bytes []byte) (*NetConf, string, error) {
...
+		curNetns = netns
+	} else {
+		curNetns, err = ns.GetCurrentNS()
+		if err != nil {
+			return nil, fmt.Errorf("failed to get init netns: %v", err)
+		}
+	}
+
```
Comment (dcbw) [inline plugins/main/vlan/vlan.go]: I feel like the code could be tightened up here, but also note that the "RandomVethName()" function must be called in the namespace that veth will be created in, so it must also be under curNetns.Do().  So the first half of the function from LinkByName(conf.Master) until LinkAdd(v) should be within a curNetns.Do() block.

File: plugins/main/vlan/vlan.go [hunk 4]
```diff
@@ -127,6 +159,10 @@ func cmdAdd(args *skel.CmdArgs) error {
 	}
 	defer netns.Close()
 
+	if n.MasterInContainer != false {
+		args.IfName = getIfName(n.Master, n.VlanId)
+	}
```
Comment (dcbw) [inline plugins/main/vlan/vlan.go]: We shouldn't be doing this here or in cmdDel().  CNI requires that the plugin use the name passed as CNI_IFNAME, and if that's eth0 then the vlan interface must be called eth0.  We can't override it from the plugin per the spec.

**Comment by rkamudhan** [comment]
@dcbw I will go through the changes and get back to you.

**Comment by jellonek** [comment]
I've contacted guys from Intel working on SR-IOV and looks like this is no longer needed by anyone.

**Pull request closed without merging**